homeassistant/components/eheimdigital/coordinator.py

```diff
@@ -0,0 +1,103 @@
+"""Data update coordinator for the EHEIM Digital integration."""
+
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+from typing import Any, NamedTuple
+
+from eheimdigital.device import EheimDigitalDevice
+from eheimdigital.hub import EheimDigitalHub
+from eheimdigital.types import EheimDeviceType
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, LOGGER
+
+PLATFORMS_PER_DEVICE_TYPE = {
+    EheimDeviceType.VERSION_EHEIM_CLASSIC_LED_CTRL_PLUS_E: [Platform.LIGHT]
+}
+
+type AsyncSetupDeviceEntitiesCallback = Callable[
+    [EheimDigitalUpdateCoordinator, str, AddEntitiesCallback], Coroutine[Any, Any, None]
+]
+
+
+class CallbackTuple(NamedTuple):
+    """A tuple for the platform setup callbacks."""
+
+    async_setup_device_entities: AsyncSetupDeviceEntitiesCallback
+    add_entities_callback: AddEntitiesCallback
+
+
+class EheimDigitalUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, EheimDigitalDevice]]
+):
+    """The EHEIM Digital data update coordinator."""
+
+    platform_callbacks: dict[Platform, CallbackTuple]
+    config_entry: ConfigEntry
+    hub: EheimDigitalHub
+    known_devices: set[str]
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the EHEIM Digital data update coordinator."""
+        super().__init__(
+            hass, LOGGER, name=DOMAIN, update_interval=DEFAULT_SCAN_INTERVAL
+        )
+        self.hub = EheimDigitalHub(
+            host=self.config_entry.data[CONF_HOST],
+            session=async_get_clientsession(hass),
+            loop=hass.loop,
+            receive_callback=self._async_receive_callback,
+            device_found_callback=self._async_device_found,
+        )
+        self.known_devices = set()
+        self.platform_callbacks = {}
+
+    def add_platform_callbacks(
+        self,
+        platform: Platform,
+        async_setup_device_entities: AsyncSetupDeviceEntitiesCallback,
+        add_entities_callback: AddEntitiesCallback,
```

Comment:
Why do we pass the async_add_entities here? I always feel dirty when doing that as you kinda lose control of where you add entities. I find it more correct to have the logic of what devices to add in the `async_setup_entry` of the platforms. This way you always have 1 file where you can see both what devices are added and when. Another solution what I would accept is that if the logic of what devices to add to what platforms is centralized, but then uses a callback to add it in the platform. I think this leaves for better understandable code instead of the parameters going into a black box

Reply:
The thing is, when `async_setup_entry` is reached there is no guarantee that all devices or even any at all (except the main device) are discovered. I need a way to add entities as the devices are found, whenever that happens. Or is there a way without needing the `async_add_entities` callback from the platforms?

Reply:
So, I have reworked it now, so that it is no longer passing `async_add_entities` around.